devicedefender/source/DeviceDefender.cpp

```diff
@@ -0,0 +1,201 @@
+
+        DeviceApiHandle::~DeviceApiHandle() { aws_iotdevice_library_clean_up(); }
+
+        void ReportTask::s_onDefenderV1TaskCancelled(void *userData)
+        {
+            auto *taskWrapper = reinterpret_cast<ReportTask *>(userData);
+            taskWrapper->m_status = ReportTaskStatus::Stopped;
+
+            if (taskWrapper->OnTaskCancelled)
+            {
+                taskWrapper->OnTaskCancelled(taskWrapper->cancellationUserdata);
+            }
+        }
+
+        ReportTask::ReportTask(
+            Aws::Crt::Allocator *allocator,
+            std::shared_ptr<Crt::Mqtt::MqttConnection> mqttConnection,
+            const Crt::String &thingName,
+            Crt::Io::EventLoopGroup &eventLoopGroup,
+            ReportFormat reportFormat,
+            uint64_t taskPeriodSeconds,
+            uint64_t networkConnectionSamplePeriodSeconds,
+            OnTaskCancelledHandler &&onCancelled,
+            void *cancellationUserdata) noexcept
+            : OnTaskCancelled(std::move(onCancelled)), cancellationUserdata(cancellationUserdata),
+              m_allocator(allocator), m_status(ReportTaskStatus::Ready),
+              m_taskConfig{mqttConnection.get()->GetUnderlyingConnection(),
+                           ByteCursorFromString(thingName),
+                           aws_event_loop_group_get_next_loop(eventLoopGroup.GetUnderlyingHandle()),
+                           reportFormat,
+                           aws_timestamp_convert(taskPeriodSeconds, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL),
+                           aws_timestamp_convert(
+                               networkConnectionSamplePeriodSeconds,
+                               AWS_TIMESTAMP_SECS,
+                               AWS_TIMESTAMP_NANOS,
+                               NULL),
+                           ReportTask::s_onDefenderV1TaskCancelled,
+                           this},
+              m_lastError(0)
+        {
+            aws_iotdevice_library_init(allocator);
+        }
+
+        ReportTask::ReportTask(ReportTask &&toMove) noexcept
+            : OnTaskCancelled(std::move(toMove.OnTaskCancelled)), cancellationUserdata(toMove.cancellationUserdata),
+              m_allocator(toMove.m_allocator), m_status(toMove.m_status), m_taskConfig(std::move(toMove.m_taskConfig)),
+              m_owningTask(toMove.m_owningTask), m_lastError(toMove.m_lastError)
+        {
+            toMove.OnTaskCancelled = nullptr;
+            toMove.cancellationUserdata = nullptr;
+            toMove.m_allocator = nullptr;
+            toMove.m_status = ReportTaskStatus::Stopped;
+            toMove.m_taskConfig = {0};
+            toMove.m_owningTask = nullptr;
+            toMove.m_lastError = AWS_ERROR_UNKNOWN;
+        }
+
+        ReportTask &ReportTask::operator=(ReportTask &&toMove) noexcept
+        {
+            OnTaskCancelled = std::move(toMove.OnTaskCancelled);
```

Comment:
check against

```
if (this != &toMove)
```

And also free existing resource of `this` before moving `toMove` to `this`.

Reply:
Thanks for that call out, I'll update this in a follow up commit.